### ISSUE TITLE
Front-Running Detection

### ISSUE THREAD

**Issue by evgenydmitriev** [inca-digital/challenge#62]
> Front-running is trading stock or any other financial asset by a broker who has inside knowledge of a future transaction that is about to affect its price substantially. A broker may also front-run based on insider knowledge that their firm is about to issue a buy or sell recommendation to clients that will almost certainly affect the price of an asset.

_[Investopedia](https://www.investopedia.com/terms/f/frontrunning.asp)_

In the crypto industry, front-running is a common occurrence, with even the biggest trading venues involved. What often starts as an in-house bot to provide additional liquidity to a centralized exchange, [sometimes turns in an illegal tool](https://medium.com/swlh/a-storm-is-brewing-over-the-largest-bitcoin-exchange-f956324b449f) for to earn money by trading ahead of their own customers.

Please comment below with ideas on detecting front-running trades based on a [stream of executed trades](https://github.com/binance/binance-spot-api-docs/blob/master/web-socket-streams.md#trade-streams). More specifically, describe an algorithm you would use to create a metric capable of automatically flagging suspicious trades. Feel free to support your ideas by adding references, datasets, graphs, and code. Comments with the best ideas will be hidden to allow others to participate. Multiple submission awards are available.

### Additional Resources

* [Challenge Program Wiki](https://github.com/inca-digital/challenge/wiki/Bounty-Program-Wiki)
* [List of available tasks](https://github.com/inca-digital/challenge/issues)
* [Inca Digital website](https://inca.digital)

Don't hesitate to ask us questions by commenting in this issue or emailing us at challenge@inca.digital.

**Comment by mjmustafaev** [comment]
In other words, are you asking the algorithm to determine, based on a WebSocket streams, whether centralized exchanges such as Binance are front running against their clients like BitMEX with its 'desk' in the Medium article?

**Comment by evgenydmitriev** [comment]
> In other words, are you asking the algorithm to determine, based on a WebSocket streams, whether centralized exchanges such as Binance are front running against their clients like BitMEX with its 'desk' in the Medium article?

It takes different forms, but you gave a great example.

**Comment by mjmustafaev** [comment]
In our stream of raw data: 
```
 {
  "e": "trade",     // Event type
  "E": 123456789,   // Event time
  "s": "BNBBTC",    // Symbol
  "t": 12345,       // Trade ID
  "p": "0.001",     // Price
  "q": "100",       // Quantity
  "b": 88,          // Buyer order ID
  "a": 50,          // Seller order ID
  "T": 123456785,   // Trade time
  "m": true,        // Is the buyer the market maker?
  "M": true         // Ignore
} 
```
followings are true:
- Each transaction has a unique `Trade ID`, `Buyer order ID` and `Seller order ID`.
- `Trade time` is the time of the transaction in milliseconds. `Event time` is the time value is inherent only to the sockets.
- If `"m": true` => the trade was initiated by the sell-side; the buy-side was already in the order book. Conversely, if `"m": false` => the trade was initiated by the buy-side; the sell-side was already in the order book.

Moreover, we use the following dictionary in this analysis:
- **Block of transaction** or _target_ – a significantly large prospective transaction that will move the price of a particular financial asset (in our case, tokens).
- **Malicious actor** or _actor_ – an agent (e.g., trader, bot, etc.) that's trying to front run.
- **Binance CEX** – BCEX.

Furthermore, note that in this case we are mainly assuming that the _target_ is on the demand side. Nonetheless, if the _target_ on the supply side, this will work similarly, _actor_ sells first to minimize his/her losses.

We have to answer two following questions first in order to finally propose a solution to the issue:
1. How can _actor_ front run?
2. What footprints can an _actor_ leave while front-running?
### How can _actor_ front run?
It's extremely hard to front run in BCEX if you're a \*general\* trader without actual insider information about the _target_, as here you can't front run with higher gas fees knowing about pending transactions in the mempool or miners can't make timestamp manipulations, since it is CEX. No miners and no gas!
Only the exchange itself has the capability to front run its clients. They have all the information about who has what to trade, even they mayn't know the exact time when their client trades their bought tokens in BCEX wallets. BCEX can predict this rather easily from the customers' transaction history, or even trigger a chain of liquidation, as they can roughly know the liquidation points of their client.  
### What footprints can an _actor_ leave while front-running?
- _Actor_ has an even number of transactions.
- The _actor_'s transaction has the same `Event type` and `Symbol` as the _target_'s transaction.
- The _actor_ possibly made a profit from the transaction (by paying a lower `Price` for the purchase of tokens than _target_ paid for the purchase from the _actor_ due the price movement caused by the _target_).
- The _actor_'s transaction has a comparable `Quantity` that don't cause a sizeable price movement.
Note: we can't track single large (_target_'s) `Quantity` based on the stream of executed trades, since there have to be multiple transactions to process it.
- The _actor_'s transaction would be amidst other \*comparable\* transactions (with a little `Price` difference) to satisfy the demand of the _target_, since it's significantly large.
- Typically, the _actor_'s transaction is at the end of this stream of the above transactions, as that is where the `Price` (slippage) is the highest.
Note: If the quantity is valid, and the _target_ is a `Market maker`, the _actor_ tries to sell at the order limit of the _target_.
- Always, by definition, if the _target_ on the demand side, there are at least two (or some other even number) trades of the _actor_ are ahead of the _target_'s transaction, possibly with a similar `Quantity`, which is identifiable through the `Trade time`.
- The _actor_ possibly tries to buy the Q<sub>a1</sub> `Quantity` that at the P<sub>a1</sub> `Price` where profit=(P<sub>t2</sub>-P<sub>a1</sub>)*Q<sub>a1</sub> is the largest when the _actor_ identifies that the _target_ is on the way. The _actor_ must take into account its impact on the price by making its first malicious transaction when choosing Q<sub>a1</sub>; and, where  P<sub>t2</sub> is the purchase `Price` of the _target_. 
Note: the transaction fees belong to the _actor_ anyway.
- The _actor_ is most likely a `Market maker` and, conversely, the _target_ is a Market taker.
### How to detect this front run?
Basically, the most unique sign of a front runner is that s/he is the one, who tries to max out the profit s/he gets from the price movement. To detect the _actor_, we need to construct a metric that calculates the probability (or some other alternative score) and flags the transactions which have a significant probability of being a front-running trade for further investigation (e.g., where we can add social media footprint layer, too).
We can build a neural network, (or some kind of linear equation to compute scores), where the nodes will be the footprints, and each footprint will have a weight depending on how close to the front-running behavior (e.g., footprints, such as _h_, which are highly unique to the front runner will have higher weight).

**Comment by mjmustafaev** [comment]
In the issue, the stream of executed trades was linked to Binance's CEX, and you mentioned CEX, too. Thus, I thought you were asking for this kind of example. Besides, the ways of detecting quite common front-running issues such as displacement, insertion (sandwich) and suppression in DEXs' DApps like Uniswap are widely available over the [Internet](https://www.researchgate.net/publication/339890096_SoK_Transparent_Dishonesty_Front-Running_Attacks_on_Blockchain). It's not quite hard when you know the (fixed) wallets the money is transferred during the run. I love to hear your feedback. We can make it better, of course!)

**Comment by DonR428** [comment]
From what I can tell, prevention of Front Running Trades seems as if most prevention is within the ability to analyze the Trades as well as any market maker trades. I think analyzing large trades if those trades are with or against the market, how many of those orders are coming from the same account. You want a code that first tags the accounts with large trades for and against the market, then you would want another filter for market makers, and frequency of trades.  In the case of Bitmex, a clear example was provided of how data analytics can at the very least show any risk of a particular company. In regards to the previous, creating a dataset for these companies, policies that allude to foul-play, server issues during high trades, frequency of trades halted due to server issues, when do server issues occur in regards to volatile trades, and so on. Once the key points are nailed down, you can begin to build a key of factors that add risk to the company. I would add one last category that compiles how many flagged risk items there are. Once that is added you would then be able to filter all the companies with a certain amount of Risk items. From there, if no improvements or changes are recommended I would suggest not recommending high risk for front running trade companies, and continually add companies to the list. This way the dataset builds more and more with each company. As a sole person alone, the dataset could be complied and put up somewhere like tableau where data can be verified and added accordingly. This way I’d have more eyes on my work and suggestions to improve this dataset. The more accurate and more ideas are added the more front-running trades we can avoid. Thank you for your time, please provide feedback if possible. Thanks again!

**Comment by Nanobelka** [comment]
Usually there is a spread between buy and sell orders, the buy price is lower than the sell price.
1. If the buy price is higher than the sell price, this is the first signal.
2. A large buy order above the sell price (or sell below the buy price) is a strengthening signal.
3. If there is a large buy order and a large sell order at the same time (and the buy price is higher than the sell price), one of the orders is probably a distraction.
4. Sometimes splitting is used to hide large orders.
5. To hide the real bid, many bids can be placed with the minimum possible number (minimum lot).
6. The order is placed as close as possible to the start of trading.
7. Sometimes over-the-counter transactions are carried out in this way at a price that is very different from the market. The order is placed a second before the start of trading.

**Comment by Kaesoron** [comment]
Front-running relies on a basic condition: clients of the trading platform must receive information with a significant delay. This will give the broker time to process the information and place orders on the exchange. 
Most likely, technically, this will be implemented as a permanent time lag in providing market information to the clients of such a broker. So, the fact of front-running can be identified by: 
1) There is a significant (and constant) time lag in the provision of market information. This can be identified by comparing the broker's trading terminal with the terminal of another, reference one. 
2) If we track the moments of mass entry / exit from positions, we will see several accounts that are ahead of must of orders by several seconds. These may be lucky traders, but in the case of automatic trading, we will observe the same time intervals, amounts and accounts. 
3) If we have access to account data, we will observe an abnormally high percentage of successful transactions and trading with the maximum available leverage from the same accounts. 
4) Such trading will be intra-day, scalping, and not a long-term strategy. When the trend reverses, the corresponding order will be placed instantly, since it makes no sense to set up trading algorithms for long-term investment. Then we will watch how brokers bid against their clients - a sell order will be placed while other clients are still placing buy orders because of time-lag. 
5) In the most severe cases, we will observe the manipulation of market information - abnormal candlesticks that will lead to automatic stopping and closing the positions of the broker's clients. This can be found out by comparing short-term charts with other brokers.

**Comment by arielchernyy** [comment]
According to my understanding, the essence of front-running is when a market participant from some source learns some information that is likely to influence the market substantially (move market prices in some direction), and buys/sells the asset to benefit from the future price movements that occur once the information becomes widely available.

The data on stream of executed trades provided by Binance contains only information on price and volume of individual trades (probably Buyer/Seller order ID’s can be used to infer Buyers/Sellers, but I’m not sure), so we have to use only this information to decide on potential front-running cases. In other words, we need to find some anomalies in the flow of trades that can be considered to be front-running episodes.

New significant information becoming publicly available should lead to relatively large price movements (that’s one of the characteristics of “significance”). So, front-running is something that happens shortly before prices move relatively large.

To truly benefit from having yet-private information, front-runners deals should probably be larger than "normal" deals. So, we are looking for “unusually large deals happening shortly before prices move unusually large”.

In the code I define periods of unusually large price movements as a coincidence of to factors: relatively large on-off price move followed by a period of unchanged prices or prices going in the same direction (to distinguish between a sustained shift in prices and usual volatility). I use the following criteria: 1 period pct_change > some cutoff AND MA_pct_change over next several periods > some cutoff.

I use second-long intervals for BTCUSD, but the code can easily be extended for any time domain.

Then, for relatively large trades, I look for trade volumes within several seconds (2 in the code, but can be modified) immediately before this unusual price move begins, that are larger than some cutoff

In the code (which is just an example using only 10 minutes of data) for cutoffs I use different percentiles of the relevant data (e.g. 0.95 or 0.975 percentiles).
If a much longer, say, weeks of months of data, time series were available, and if we had a set of confirmed front-runners deals, all the hyper-parameters could be fine-tuned (via grid search or genetic algorithm of some kind) to find optimal values resulting in the largest share of front-runners’ deals detected

https://github.com/arielchernyy/FrontRunning/blob/daff45d5a754c7037e19a3582819a64b3c1d93c3/front_running_ArielChernyy.ipynb

**Comment by evgenydmitriev** [comment]
The challenge is still open. Many of the challenge participants focused on investigative approaches that involved manual analysis of specific cases of front-running. This, however, is an engineering challenge, requiring successful submissions to include an algorithm, supported by references, datasets, graphs, and/or code. We have more than enough of ingenious ideas on how it can be done, but no solid plans of how to implement it using real-time streaming data.

**Comment by mdrbase** [comment]
Im sure people are already doing this but if anyone is so inclined im sure they could create a "honeypot" to either identify and track   or potentially deceive people who are attempting.
